File: libs/power-log-parser/src/lib/parsers/blood-gem-buff-changed-parser.ts
import { GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Node, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class BloodGemBuffChangedParser implements ActionParser {
	readonly ParserName = 'BloodGemBuffChangedParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, facade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = facade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === TagChange &&
			((node.Object as TagChange).Name === (GameTag.BACON_BLOODGEMBUFFATKVALUE as number) ||
				(node.Object as TagChange).Name === (GameTag.BACON_BLOODGEMBUFFHEALTHVALUE as number))
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const entity = this.GameState.CurrentEntities.get(tagChange.Entity);
		if (entity == null) {
			return null;
		}

		const controller = entity.GetController();
		if (controller !== this.StateFacade.LocalPlayer!.PlayerId) {
			return null;
		}

		const atk =
			tagChange.Name === (GameTag.BACON_BLOODGEMBUFFATKVALUE as number)
				? tagChange.Value
				: entity.GetTag(GameTag.BACON_BLOODGEMBUFFATKVALUE, 0);
		const health =
			tagChange.Name === (GameTag.BACON_BLOODGEMBUFFHEALTHVALUE as number)
				? tagChange.Value
				: entity.GetTag(GameTag.BACON_BLOODGEMBUFFHEALTHVALUE, 0);
		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'BLOOD_GEM_BUFF_CHANGED',
				GameEventHelper.CreateProvider(
					'BLOOD_GEM_BUFF_CHANGED',
					null as any,
					entity.GetEffectiveController(),
					entity.Id,
					this.StateFacade,
					{
						Attack: atk,
						Health: health,
					},
				),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/BloodGemBuffChangedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class BloodGemBuffChangedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public BloodGemBuffChangedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && ((node.Object as TagChange).Name == (int)GameTag.BACON_BLOODGEMBUFFATKVALUE 
                    || (node.Object as TagChange).Name == (int)GameTag.BACON_BLOODGEMBUFFHEALTHVALUE) ;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var entity = GameState.CurrentEntities.GetValueOrDefault(tagChange.Entity);
            if (entity == null)
            {
                return null;
            }

            var controller = entity.GetController();
            if (controller != StateFacade.LocalPlayer.PlayerId)
            {
                return null;
            }

            var atk = tagChange.Name == (int)GameTag.BACON_BLOODGEMBUFFATKVALUE ? tagChange.Value : entity.GetTag(GameTag.BACON_BLOODGEMBUFFATKVALUE, 0);
            var health = tagChange.Name == (int)GameTag.BACON_BLOODGEMBUFFHEALTHVALUE ? tagChange.Value : entity.GetTag(GameTag.BACON_BLOODGEMBUFFHEALTHVALUE, 0);
            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                 "BLOOD_GEM_BUFF_CHANGED",
                GameEvent.CreateProvider(
                    "BLOOD_GEM_BUFF_CHANGED",
                    null,
                    entity.GetEffectiveController(),
                    entity.Id,
                    StateFacade,
                    //null,
                    new {
                        Attack = atk,
                        Health = health,
                    }),
                true,
                node) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
